src/outputs/restart.hpp

```diff
@@ -83,262 +64,62 @@ class RestartReader {
     }
   };
 
-  SparseInfo GetSparseInfo() const;
-
-  // Return output format version number. Return -1 if not existent.
-  int GetOutputFormatVersion() const;
+  [[nodiscard]] virtual SparseInfo GetSparseInfo() const = 0;
 
- private:
-  struct DatasetHandle {
-    hid_t type;
-    H5D dataset;
-    H5S dataspace;
-    int rank;
-    hsize_t count;
-    std::vector<hsize_t> dims;
+  struct MeshInfo {
+    int nbnew, nbdel, nbtotal, root_level, includes_ghost, n_ghost;
+    std::vector<std::string> bound_cond;
+    std::vector<int> block_size;
+    std::vector<Real> grid_dim;
+    std::vector<int64_t> lx123;
+    std::vector<int> level_gid_lid_cnghost_gflag; // what's this?!
   };
+  [[nodiscard]] virtual MeshInfo GetMeshInfo() const = 0;
 
-  // internal convenience function to open a dataset, perform some checks, and get
-  // dimensions
-  template <typename T>
-  DatasetHandle OpenDataset(const std::string &name) const {
-#ifndef ENABLE_HDF5
-    PARTHENON_FAIL("Restart functionality is not available because HDF5 is disabled");
-#else  // HDF5 enabled
-    DatasetHandle handle;
-
-    // make sure dataset exists
-    auto status = PARTHENON_HDF5_CHECK(H5Oexists_by_name(fh_, name.c_str(), H5P_DEFAULT));
-    PARTHENON_REQUIRE_THROWS(
-        status > 0, "Dataset '" + name + "' does not exist in HDF5 file " + filename_);
-
-    // open dataset
-    handle.dataset = H5D::FromHIDCheck(H5Dopen2(fh_, name.c_str(), H5P_DEFAULT));
-    handle.dataspace = H5S::FromHIDCheck(H5Dget_space(handle.dataset));
-
-    // get the HDF5 type from the template parameter and make sure it matches the dataset
-    // type
-    T *typepointer = nullptr;
-    handle.type = getHDF5Type(typepointer);
-    const H5T dset_type = H5T::FromHIDCheck(H5Dget_type(handle.dataset));
-    status = PARTHENON_HDF5_CHECK(H5Tequal(handle.type, dset_type));
-    PARTHENON_REQUIRE_THROWS(status > 0, "Type mismatch for dataset " + name);
+  struct TimeInfo {
+    Real time, dt;
+    int ncycle;
+  };
```

Comment:
Same for this. Don't we already have an appropriate struct for this actually?

Reply:
All/most these variables live as member of `Mesh`.
Here,they are being used only for restarts, i.e., these are temp structures that contain the necessary info that is being *read* from the restart file and then passed to the appropriate places in the code allowing to hide the backend specific `GetAttr<T>` calls.
This is mirroring the existing `SparseInfo` struct used for restarts and I split this into `Time` and `Mesh` from a logical point of view.

Reply:
Done.